### PR TITLE
Attempt to quote anyway if unquoting fails

### LINKED ISSUE #2356: Requests unable to follow/retrieve links with percent in url
A simple requests.get(url) doesn't work for the following:

http://bit.ly/1x5vKWM
http://bit.ly/1yPgqvg
http://style.shoedazzle.com/dmg/3AE3B8?dzcode=FBT&dzcontent=FBT_SDZ_CPM_Q414&pid=112768085&aid=285880402&cid=0&publisher=%ppublisher=!;&placement=%pplacement=!;

### PR DESCRIPTION
**Pull request by sigmavirus24**
Fixes #2356

TODO

- [x] Add test

### CODE REVIEW

**Comment by Lukasa** [comment]
Looks fine, modulo a test. =)

**Comment by kevinburke** [comment]
where does the list of safe characters come from? maybe this could be pulled into a named var for clarity?

**Comment by sigmavirus24** [comment]
## It comes from the RFCs

Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by kennethreitz** [comment]
Update?

**Comment by sigmavirus24** [comment]
Update: forgotten. I've been really busy lately but I'll prioritize this. Sorry for losing track of it.

**Comment by sigmavirus24** [comment]
Added two tests and gave the strings slightly better names. Not sure we need to spell out every last RFC referenced in the comments though since they're all pretty well known. @Lukasa I'd appreciate some review.

File: requests/utils.py [hunk 1]
```diff
@@ -418,10 +418,18 @@ def requote_uri(uri):
...
+    safe_with_percent = "!#$%&'()*+,/:;=?@[]~"
+    safe_without_percent = "!#$&'()*+,/:;=?@[]~"
+    try:
+        # Unquote only the unreserved characters
+        # Then quote only illegal characters (do not quote reserved,
+        # unreserved, or '%')
+        return quote(unquote_unreserved(uri), safe=safe_with_percent)
+    except InvalidURL:
+        # We couldn't unquote the given URI, so let's try quoting it, but
+        # unquoted '%'s in the URI. We need to make sure they're properly
```
Comment (Lukasa) [inline requests/utils.py]: This part of the comment is gibberish. ;)

File: test_requests.py [hunk 2]
```diff
@@ -1301,6 +1301,28 @@ def test_get_auth_from_url(self):
...
+
+        See: https://github.com/kennethreitz/requests/issues/2356
+        """
+        from requests.utils import requote_uri
+        bad_uri = ('http://style.shoedazzle.com/dmg/3AE3B8?dzcode=FBT'
+                   '&dzcontent=FBT_SDZ_CPM_Q414&pid=112768085&aid=285880402'
+                   '&cid=0&publisher=%ppublisher=!;&placement=%pplacement=!;')
+        quoted = ('http://style.shoedazzle.com/dmg/3AE3B8?dzcode=FBT'
+                  '&dzcontent=FBT_SDZ_CPM_Q414&pid=112768085&aid=285880402&cid'
+                  '=0&publisher=%25ppublisher=!;&placement=%25pplacement=!;')
```
Comment (Lukasa) [inline test_requests.py]: These URIs are massive so it makes it easy to miss what you're actually doing. Can we shorten them to a smaller case?

**Comment by Lukasa** [comment]
Some small notes. =)

**Comment by sigmavirus24** [comment]
@Lukasa forgot to mention this was updated to address your feedback.

**Comment by Lukasa** [comment]
LGTM. :+1:

**Pull request merged**